### ISSUE TITLE
Guidelines on adding new icons

### ISSUE THREAD

**Issue by matthijsmelissen** [gravitystorm/openstreetmap-carto#660]
There are quite a few requests for adding new icons (#591 #540 #570 #487 #456 #452 #450 #419 #418 #336 etc), and even more features that we currently do not render, but might be nice to have on the map.

How do decide which POIs we render as icon?

Do we want to increase the icon set, or keep the set of icons small?

@gravitystorm you indicated in the past that you are in favour of an approach that doesn't involve hundreds of new icons. Could you give some more background to that statement?

A large icon set might give more information. But having too many icons might also make the map unreadable. We should also make sure that all icons are understandable.

**Comment by brycenesbitt** [comment]
I think it's time to reframe this debate.

POI's, particularly obscure ones, don't clutter the map.
Line styles and fills get confusing quickly.

As such I think the "1000 uses" threshold is particularly lacking in vision.  It's an electronic map: it can support many constituencies, including small ones.

**Comment by imagico** [comment]
As @math1985 said the main purpose of displaying POI icons is to allow map users to search for a particular type of feature.  This only works if there is a limited set of symbols, if there are thousands of different icons it does not (needle in a haystack problem).  With the brown icons this mostly still works, with the purple ones it no more does in most cases (IMO).

**Comment by matkoniecz** [comment]
> main purpose of displaying POI icons is to allow map users to search for a particular type of feature

Ad second purpose is too recognize object based on icon and context (IMHO works for shop icons like ones added in #1493 and is not working so well in case of some other icons like #1776)

**Comment by brycenesbitt** [comment]
My main purpose for POI icons is to see what's in an area.  I rarely search.
OSM is one dataset, but many many uses.

I have no needle in the haystack problem with a multitude of POI types.
Rather the opposite.
If anyone does have a needle in the haystack problem, that's a problem with
the client.  It should in search show the most commonly needed
options, with a pathway to getting more.

**Comment by kocio-pl** [comment]
I generally think the same as @brycenesbitt - I don't look for exact objects, I rather want to know what's up there (as much as it's possible) and I think something is wrong when there are holes or just some unidentified structures which don't look like plain buildings. POIs are helping me understand the area along with roads, land colors, names and boundaries. The higher zoom, the more important POIs and icons are.

@imagico 
> This only works if there is a limited set of symbols, if there are thousands of different icons it does not (needle in a haystack problem).

The map areas and lines also would work better with limited set of road types or land colors, so this argument is universal and works not only for POIs/icons. After all this is the same haystack for every item - try to identify anything other than the icon symbols (it starts with lines description):

http://wiki.openstreetmap.org/wiki/Standard_tile_layer#Lines

and tell if it's any easier for you - for me it's not.

@matkoniecz 
> And second purpose is too recognize object based on icon and context (IMHO works for shop icons like ones added in #1493 and is not working so well in case of some other icons like #1776)

Which leads me to a question "how does it work now?":

1) How do we know the basket icon means convenience shop and not the:
- general shop (that's how it's used in HOT)
- basket shop (I know at least one such place)
- self-service shop (like in some consumer electronics for example)?

2) How do we recognize that given area is:
- allotments
- scree/shingle/bare rock
- heath
- orchard
- vineyard (with new pattern)?

For me all these examples show how much arbitrary some objects are - yet we still use them and it would be hard to find anything better. Of course some symbols/colors are easier: that's why new tree area rendering is better than the old one and why I created new pattern for quarry, we also are trained to recognize the railway, water or churches without any additional clues.

But does it mean we should avoid such generalizations and stop rendering all these "muddy" examples just because there's no way to tell without looking at the cheat sheet what the hell is this green area (other than that it's probably about grass; or some other plants; or the sport; or leisure)? And if we shouldn't - what are the reasons?

BTW: if you don't like the icon for social facilities and want to stick with the name, I'm curious:
- how the lack of general hint ("people together" icon -> something social, community), which gives you less context, makes it easier for you to recognize the object?
- how do you plan to recognize such places more accurate in the languages you don't know at all?

I know these question are tricky, but I think the whole subject is not that easy and that's why I try to be as concrete as possible.

**Comment by imagico** [comment]
> try to identify anything other than the icon symbols (it starts with lines description): http://wiki.openstreetmap.org/wiki/Standard_tile_layer#Lines and tell if it's any easier for you - for me it's not.

Are you kidding?

The line features use a huge variety of different colors, line patterns and combination and more importantly recognizing them has as much to do with the geometry of the line and context as it has with the styling of the lines themselves.  The icons OTOH are shown in one of very few color and are always identical.

If you don't believe me just make a test: try shuffling all the shop, food/drink and culture icons in a map and see if it still looks plausible at a quick glance and then try shuffling all the highway and boundary line types.

Compared to other maps the number of fixed design point symbols is already very large in this style.  There are good reasons why point symbols are used relatively sparsely in many cartographic designs - they take a lot of space, transport relatively little information and frequently obscure other, more meaningful features which often happen to be located around where the icon is.

**Comment by brycenesbitt** [comment]
@imagico from my point, I'm not kidding. There are far too many line styles and colors for me to keep track of.  Show me one in isolation and I won't be able to tell you what it means.

The lines exist in a context.  But so do the icons.  A "toilet holding tank dump station" icon in a campground is in context.  A bicycle repair stand near bike racks is in context.  A road's name gives it context.  A POI's name gives it context.  And it's an _electronic map_, should a user be confused the back end support exists to allow a click or swipe to resolve the confusion.

And consider how the "1,000 uses in the database rule" promotes clutter.  A common but low utility POI rates higher than a rare but crucial POI.

**Comment by kocio-pl** [comment]
> The line features use a huge variety of different colors, line patterns and combination and more importantly recognizing them has as much to do with the geometry of the line and context as it has with the styling of the lines themselves. 

The ultimate context for those lines colors and other features is exactly this "haystack". If you don't believe, try to tell if it's primary or tertiary road with no cheat sheet. Oh - you already know it probably, so the test will fail. But looking just at the map you can also tell "it's probably more important road than the other" or "I guess this road (track) is more solid than the other one", but nothing as exact as you may like. Colors and other elements are not tied to some road features (they are just black, gray or brown in reality), that's why it's possible at all to try a new set.

> The icons OTOH are shown in one of very few color and are always identical.

I don't get how the icons are identical to you. They have just similar size (and visual style, if we were able to achieve it! It was great that @nebulon42 started the process of unifying them in this regard). Take just two random icons and if you really see no difference, I think it's rather you who is kidding me...

> Compared to other maps the number of fixed design point symbols is already very large in this style. 

Because this is how it is in reality. There are many important kinds of "points" (the icons are also used for areas) in the world and we're able to catch and show them, which is great and unique property of OSM!

>There are good reasons why point symbols are used relatively sparsely in many cartographic designs - they take a lot of space, transport relatively little information and frequently obscure other, more meaningful features which often happen to be located around where the icon is.

This is very important point, so let's look closely: if the map has low-to-medium zoom level, you have to flush away a lot of things and stick mainly to the areas in the low zoom and lines in the medium zoom. This is not to say that areas are not important on high zoom level (like water) or that there are no important points in the low zoom level (capitals), but you get the pattern.

On the high zoom level the world is much different. Think about what you see with a telescope, a naked eye - and the microscope. You can see no continents now (just a few of them exist), there are only few lines like streets (there are at most a few dozen types of them) and probably few more points - but only in the given box, because there are a lot of them if you add all the boxes.

So - at the high zoom levels (where most of the icons exist!) all these reasons you mentioned are no longer valid. Most of the time you have plenty of rather empty space - no matter how crowded place you'll examine! There's not much to obscure, because points typically don't stack like landuses, for example. And "relatively little information" is, well, really relative - here this is the most interesting thing you can get! We're probably not able to really show the opening hours on any static piece of map with zmax=19, but the name and color are just too general, so the icon is here to tell us more about specific type of this point in a common, uniform way.

For me the reasons you just gave are perfect example of the "scale blindness". When you know what typical map should be, you just seem to ignore the reality of high zoom level maps - which are very different. I'm all for the rules you gave (they are sane and useful), but only when dealing with lower zoom levels - they're unusable outside this scope. Flushing things away is necessary when you have to do it, but it's just wasting the precious informations if you don't. And we have of lot of them in OSM.

The last thing - you're not supposed to know all the icons on the map without additional help if there will be a lot of them. But you're also not supposed to understand all the names in different languages and even know some alphabets we render on the map. How does it limits your ability to use the map? This is the same question which I have asked Mateusz: "how do you plan to recognize such places more accurate in the languages you don't know at all?" (if not using visual symbols like icons).

OSM is not just "my home yard", the world is much bigger and if we want to have as universal map style as it's possible, icons will help us. Let's not be "blind" also for this aspect of scale.

**Comment by dieterdreist** [comment]
sent from a phone

> Am 26.08.2015 um 18:54 schrieb brycenesbitt <notifications@github.com>:
> 
> The lines exist in a context. But so do the icons. A "toilet holding tank dump station" icon in a campground is in context. A bicycle repair stand near bike racks is in context. A road's name gives it context. A POI's name gives it context. And it's an electronic map, should a user be confused the back end support exists to allow a click or swipe to resolve the confusion.


I agree for very high zooms (maybe z19 and higher). Christoph is right for z17 maybe 18 and lower. In very detailed close ups it would indeed be useful to have an icon for "sub-features" that occur inside other features

**Comment by kocio-pl** [comment]
My idea is rather gradually showing them more or less by their size: 
- z17 for normal shops and standard size amenities like banks, post offices and so on (as it is now in general),
- z18 for smaller ones (phones, ATMs, post boxes, recycling bins, probably vending machines too - I plan to move them here)
- z19 for really small, less important in most cases and very  (as it is now: benches, small waste bins, but also emergency phones or AED, because they're just in case).

**Comment by brycenesbitt** [comment]
Some composite icons are possible also.

For example campgrounds may have attributes of fee, drinking_water, toilet
dump station.
In a real campground you may see a notice board showing a large tent symbol
with a series of
small symbols for what's available (e.g. toilets / beach / wildlife
viewing).


Such an organization gives instant context, and exposes data that presently
is not rendered at any zoom.

**Comment by imagico** [comment]
> Some composite icons are possible also.

This is something that can work very well if done well - this is why i wrote *fixed design point symbols* above - if you make minor variations and additions to the same base shape this is not as cluttering as fully different icons.

The key here is to make the base shape really recognizable.  There are a number of recurring base shapes in the current icon set like the car and the bed shape but they occur in very different sizes and variants so are not really recognizable as the same base feature type (and are also often not meant as such).

This technique also usually works best with fairly simple and abstract shapes.  Good example are medical symbols - cross within circle, cross within box etc. Such usually require a legend to be understood but once the map user is familiar with them they are very well readable.

**Comment by kocio-pl** [comment]
Having designed a bunch of icons myself, I think currently we're too constrained technically: 14x14 px is barely enough for one/two symbols (that's why I admire department_store icon!). I guess composite icons could be good on the additional layer (given we have more pixels there), but it's outside the scope of this style.

**Comment by brycenesbitt** [comment]
Just keep in mind the icon does not do all the "heavy lifting", not when
there's a text label.
I agree fully that 14x14px is a terrible constraint, and with high DPI
desktop displays, really a historic artifact.

**Comment by nebulon42** [comment]
14px is great because it prevents you from squeezing in too much. In that vein I think the department store is horrible, although I have so far no better alternative to propose. Note that the 14px where no magic number I came up with, I started with 16px icons and wanted to have 1px padding. Sometimes 14px is a little tricky and if it would be now 16px or 18px, why not. But if you recall the discussion on icon halos (which I still think are really necessary for legibility) there was a lot of concern that 18px is too much.

If you look to Maki they currently even have smaller versions that are used by e.g. iD to display marker symbols. So there could be some need for small icons. Regarding high res displays it is very easy to make icons larger (e.g. double size), but downsizing an icon should be avoided if you don't want to end up with blurry icons. Since we don't serve retina tiles yet (do we?) these constraints are far from becoming historical.

@imagico It would be great if you would be a bit more specific where you see inconsistencies in current shapes. Regarding the car example the only inconsistency I see is the car shop icon. I refrained from changing that because that was already SVG and didn't want to give the impression that wanted to change everything to "my" icons.

**Comment by imagico** [comment]
I did not see particular inconsistencies but there are currently relatively few icons that are clearly variants of the same base icon (in contrast to composite icons built from the same base elements).  Variants in this sense means the dominating element of the icon is the same in all variants.

The clearest case IMO is shelter/alpine hut.  Parking/bicycle parking as well although the 'P' has significantly different proportions in both (the smaller one is notably thicker in lines compared to the character size).  In most other symbols with recurring elements the recurring element is not clearly dominating so the effect is different.

**Comment by kocio-pl** [comment]
Sounds like maybe scaling down the standard (=car) parking icon to be the same as the bicycle/motorcycle would be good indeed. We could also add a car icon to make them more consistent, but I'm not sure if it won't be too strong given how popular car parkings are (600k:60k:60).

**Comment by nebulon42** [comment]
@imagico Got your point now. I was thinking of how to best design the "progression" from weather shelter via basic shelter towards alpine hut and onwards to guest house and the like for some time now. So far I have not found a working solution. Alpine hut is not bad, but the hiker is far too detailed for me.

If the motorcycle parking icon was derived from the bicycle parking icon it might be a better idea to upscale that to the general parking icon. In my repository there is also a general parking icon, but I don't remember currently if it is similar to the bicycle parking one. Generally, I always try to reuse shapes as much as possible, but sharpness is - at least for me - an equally important constraint.

**Comment by kocio-pl** [comment]
@nebulon42 
> In that vein I think the department store is horrible, although I have so far no better alternative to propose.

I admire this icon simply because it managed to squeeze so much things, yet I can read it easily. But I don't think it couldn't be made in a different way. Maybe you could try with two other left/right halves (half the basket, half the shirt for example)?

> I refrained from changing that because that was already SVG and didn't want to give the impression that I wanted to change everything to "my" icons.

Just my position: I don't have a problem with the authorship - if I like it more than the previous one, it's just better, but if I think it's not readable enough (this is the only thing which I don't like in some of your drafts), I will try to propose something different or vote the change down. In general you gave this style a lot of visual consistency, and that's what I like here a lot. And sometimes just your fresh idea is what helps to make a progress.

In a nutshell: don't be shy (unless you just don't like to discuss another item to death :wink:).

**Comment by math1985** [comment]
I have been thinking a bit more about the principles behind this. I think there are about three uses of icons in cartography. With each of the uses, I have indicated some questions map users might have, and how the map could answer them.

* As a target. The icon is used for the user to find a particular (type of) place.
** Where can I find the hospital where my friend is admitted? - At the South railway station.
** Where can I buy bread? - There is a bakery around the corner.
** Where can I get dinner? - There is an area with restaurants to the North, and a single restaurant to the South.
* As a landmark. The icon is used to aid the user in navigation.
** How do I get to Oak Street? - Turn left at the bakery.
** Where is my rental apartment? - On top of the estate agent.
** Where will the meeting take place? - Not far from the famous cathedral.
* As an area description. The icon is used to highlight the important features of an area.
** What kind of area is my hotel in? - A retail area.
** What do I see on my way to the museum? Two churches and a windmill.

In addition, in the context of OSM, it is also important to be able to check whether a particular feature has been mapped already.

All these uses are probably somewhat conflicting. For example, schools will not be often used as a target, but as a landmark they can be quite handy. Probably (and hopefully) there also lead to common requirements. 

In my opinion, our map is not that useful with using icons as targets. Because there are so many different types of icons, it is really hard to find, for example, a bakery in a fully mapped city centre. On the other hand, because a user might have so many different types of targets, it is also not that easy to get this right on a static map. We could improve on this by dropping icons for features that users are less likely to search for, such as mobile phone stores, and maybe even clothes stores etc. Also a smart use of colour could contribute. Because of maps with dynamic icons and search functions (in OsmAnd), using icons as targets will likely become less important in the future.

I think we are doing slightly better with using icons as landmarks. Up to z16, we only render icons for things that tend to be more important as landmarks on a larger scale, such as churches, museums, and theaters. We could still improve on this by not rendering very minor museums and theaters based on area size on z16. On z17, icons in big cities are not so useful as landmarks, because there is just rendered too much, and it is impossible to tell what is important. On this zoom level, for this purpose it might make sense to drop some types of icons. I would say that from z18 on, things get a bit better again, because at that scale, any shop can serve as a landmark.

I think we are also not too bad with using icons as area descriptors. One of the problems for this use, is that icons like mailboxes, water taps, etc. are unnecessary on z17 and maybe even z18.

For the mapper feedback loop, it would be useful to add even more features. For example, public_buildings, artworks, and social_facility's are currently not rendered at all. However, rendering them only at (very) high zoomlevels would suffice.
 
Probably other/smarter people have been thinking about this before me, does anyone have a recommendation where I can read more about this topic?

**Comment by kocio-pl** [comment]
Great analysis, thanks!

In general I support your observations. I also think we're not able to act effectively in terms of "targets" - we should give it up to the other, dynamic tools. I'm not sure what is the difference between "landmark" and "description" - I guess it may be about easy to spot objects (landmarks) vs general properties and both deserve our attention.

I was thinking lately about dropping some objects from z17, which tends to be the most busy level and [causing most of the discussions](https://github.com/gravitystorm/openstreetmap-carto/issues/1560). The easiest action would be to demote specific shop icons one level down and instead show just the dots on z17. This would be consistent with my idea of using z18 for smaller amenities and z19 for "the rest", because just dropping shops completely from z17 would not allow us to show the scale more accurately. 

Yet my idea of having some system was [neglected](https://github.com/gravitystorm/openstreetmap-carto/pull/1884#issuecomment-144378212) by @imagico and I have no idea what to do with it. I wanted to use a typical size as the most neutral measure I could think of, with importance as a "modifier", because it's more subjective and harder to rank, but unreasonable to completely avoid it. So we're stuck with it currently, because no other system was proposed, just a subjective list of important objects, which unfortunately is completely different.

I could also think of of adding more features in z18 and z19, which are rather safe from the danger of being overloaded, to satisfy mappers' needs, but for example @matkoniecz says he just doesn't like having more icons (no matter what the zoom level). So while I have an idea how to render artworks (which can be landmarks in parks for example), I don't know if it won't be just wasting my time trying and I'm stuck here too. The same for education icons, which are now ready, but I had some technical problems rendering them properly, so I gave up for the same reasons.

In my opinion lack of any system is the most important obstacle to do adding anything now. It looks like the coding side makes it also hard (more code lines and [performance issues](https://github.com/gravitystorm/openstreetmap-carto/issues/1941), which was reported by @pnorman lately). Some problems with icons could be solved by taking the context into account, but it also [was rather not welcomed](https://github.com/gravitystorm/openstreetmap-carto/issues/1957) because of expected complexity.

So, I share your vision, but there are different practical problems I don't know how to solve.

**Comment by Tomasz-W** [comment]
We have detailed discussion about each icon in certain tickets. There is also a guideline in `contributing.md`   https://github.com/gravitystorm/openstreetmap-carto/blob/master/CONTRIBUTING.md#map-icon-guidelines
There haven't been new comments here for more than 2 years, so maybe this issue should be closed then?

**Comment by kocio-pl** [comment]
It can be extended to cover more problems from original report, but let's reopen it then - or open more specific discussion, since guidelines is strictly related to discussing policies.

**Comment by kocio-pl** [comment]
BTW: there's a set of raster icons from (ex-)Mapzen to help designing new shapes:
https://github.com/tangrams/icons
